component/loki/source/gelf/component.go

```diff
@@ -0,0 +1,147 @@
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+var _ component.Component = (*Component)(nil)
+
+// Component is a receiver for graylog formatted log files.
+type Component struct {
+	mut       sync.RWMutex
+	target    *target.Target
+	o         component.Options
+	metrics   *target.Metrics
+	handler   *handler
+	receivers []loki.LogsReceiver
+}
+
+// Run starts the component.
+func (c *Component) Run(ctx context.Context) error {
+	defer func() {
+		c.target.Stop()
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.handler.c:
+			c.mut.RLock()
+			lokiEntry := loki.Entry{
+				Labels: entry.Labels,
+				Entry:  entry.Entry,
+			}
+			if lokiEntry.Labels["job"] == "" {
+				lokiEntry.Labels["job"] = model.LabelValue(c.o.ID)
+			}
+			for _, r := range c.receivers {
+				r <- lokiEntry
+			}
+			c.mut.RUnlock()
+		}
+	}
+}
+
+// Update updates the fields of the component.
+func (c *Component) Update(args component.Arguments) error {
+	newArgs := args.(Arguments)
+
+	c.mut.Lock()
+	defer c.mut.Unlock()
+
+	if c.target != nil {
+		c.target.Stop()
+	}
+	c.receivers = newArgs.Receivers
+	rcs := flow_relabel.ComponentToPromRelabelConfigs(newArgs.RelabelConfigs)
+	t, err := target.NewTarget(c.metrics, c.o.Logger, c.handler, rcs, convertConfig(newArgs))
+	if err != nil {
+		return err
+	}
+	c.target = t
+	return nil
+}
+
+// Arguments are the arguments for the component.
+type Arguments struct {
+	// ListenAddress only supports UDP.
+	ListenAddress        string                 `river:"listen_address,attr,optional"`
+	UseIncomingTimestamp bool                   `river:"use_incoming_timestamp,attr,optional"`
+	RelabelConfigs       []*flow_relabel.Config `river:"rule,block,optional"`
+	Receivers            []loki.LogsReceiver    `river:"forward_to,attr"`
+}
+
+func defaultArgs() Arguments {
+	return Arguments{
+		ListenAddress:        "0.0.0.0:12201",
+		UseIncomingTimestamp: false,
+	}
+}
+
+// UnmarshalRiver implements river.Unmarshaler.
+func (r *Arguments) UnmarshalRiver(f func(v interface{}) error) error {
+	*r = defaultArgs()
+
+	type arguments Arguments
+	if err := f((*arguments)(r)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func convertConfig(a Arguments) *scrapeconfig.GelfTargetConfig {
+	return &scrapeconfig.GelfTargetConfig{
+		ListenAddress:        a.ListenAddress,
+		Labels:               nil,
+		UseIncomingTimestamp: a.UseIncomingTimestamp,
+	}
+}
+
+// New creates a new gelf component.
+func New(o component.Options, args Arguments) (component.Component, error) {
+	metrics := target.NewMetrics(o.Registerer)
+	c := &Component{
+		o:       o,
+		metrics: metrics,
+		handler: &handler{c: make(chan api.Entry)},
+	}
+	// Call to Update() to start readers and set receivers once at the start.
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+type handler struct {
+	c chan api.Entry
+}
+
+func (h *handler) Chan() chan<- api.Entry {
```

Comment:
I think we can use our own`loki.Entry` and `loki.NewEntryHandler(c.handler, func() {})` directly here instead of the Promtail package. WDYT?